### DISCUSSION TITLE
Question: transact() vs build_transaction()/sign_transaction()

### DISCUSSION THREAD

**Discussion by fulltimecrypto** [smartcontractkit/full-blockchain-solidity-course-py#1566] [Q&A] [answered]
Hi there,

I find myself confused about the difference between two methods to deploy an instance of a contract: 

- Method 1 (and how it's done in the course):

 1. Build transaction using build_transaction()
 2. Sign transaction using sign_transaction() and the private key
 3. Send the transaction using send_raw_transaction() and waiting for the receipt 

- Method 2 (from the official docs https://web3py.readthedocs.io/en/latest/overview.html?highlight=transact()#contracts and other turorials such as https://www.dappuniversity.com/articles/web3-py-intro):

 1. Call transact() on the constructor method of the contract and waiting for the receipt

Both methods work perfectly fine in the SimpleStorage deployment example. However, I cannot understand why the second method works as well, without signing any transaction and not using the private key? When should one use which method and why? 

I found a similar question in the ethereum stackexchange: https://ethereum.stackexchange.com/questions/111893/whats-the-difference-between-transact-and-buildtransaction However, I'm not satisfied with the answer, since I never had to sign any transaction to deploy the contract. 

Would be very thankful for any answer that would clear my mind :)

**Comment by cromewar** [comment, marked as answer]
Hello @fulltimecrypto 
Nice question, the answer you got on stack exchange is actually correct, but let me go in deep in to it.
Using `buildTransaction` is more a step to step process, Patrick used this approach to teach how all the process to create and sign transaction works, including you have direct control of the private_key, which is a variable you can use on the code or as an environment variable.

The `transact` function requires the private key to be handled differently, most likely by the rpc client (metamask, ganache and so on).

At the end both ways end up using `send_raw_transaction`, but transact does it more automatically.

So yes, the only difference here is how the private key is managed, anyways on the further lesson you'll find that does no matter that much as we use frameworks like brownie which handles it differently.

Cheers.

**Reply by fulltimecrypto** [reply]
Hi cromewar,

Really appreciate your answer! Still I'm a bit confused as to how it's possible to deploy a contract without manually signing it. But, as you've pointed out, it seems as if this is done in the background by ganache in this case. 
Also, you are right, now that I'm more advanced in the lesson and learning how to work with brownie, the question isn't as relevant anymore. 

Thanks again.
